### ISSUE TITLE
Inconsistency of `format_args!` between clippy and rustc

### ISSUE THREAD

**Issue by oxalica** [rust-lang/rust-clippy#16736]
### Summary

I'm refactoring my code to leverage store-able `format_args` (https://github.com/rust-lang/rust/pull/140748) which is stable since 1.89, but found that `cargo clippy` sometimes reports lifetime errors where `cargo check` or `cargo build` compiles fine without any error.

The following code compiles since Rust 1.89, but still emits lifetime errors on clippy since 1.89. The latest stable and nightly clippy still fail to compile it.

### Reproducer

Code ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=eec40b7e18498407f7ebdebe3c1b4d89)):

```rust
use std::fmt;

// This function compiles fine on both rustc and clippy.
fn do_fmt1(w: &mut dyn fmt::Write, mode: u8) -> fmt::Result {
    let f = match mode {
        0 => format_args!("foo"),
        1 => format_args!("bar {}", 42),
        _ => unreachable!(),
    };
    w.write_fmt(f)
}

fn do_fmt2(w: &mut dyn fmt::Write, mode: u8) -> fmt::Result {
    w.write_fmt(match mode {
        0 => format_args!("foo"),
        1 => format_args!("bar {}", 42), // <- clippy reports lifetime error, but rustc compiles fine.
        _ => unreachable!(),
    })
}

fn main() {}
```

Current output:

```
  Checking playground v0.0.1 (/playground)
error[E0716]: temporary value dropped while borrowed
  --> src/main.rs:15:14
   |
13 |     w.write_fmt(match mode {
   |       --------- borrow later used by call
14 |         0 => format_args!("foo"),
15 |         1 => format_args!("bar {}", 42),
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^-
   |              |                        |
   |              |                        temporary value is freed at the end of this statement
   |              creates a temporary value which is freed while still in use
   |
   = note: consider using a `let` binding to create a longer lived value

For more information about this error, try `rustc --explain E0716`.
error: could not compile `playground` (bin "playground") due to 1 previous error
```

Desired output:

No error.

### Version

```text
rustc 1.94.0 (4a4ef493e 2026-03-02)
binary: rustc
commit-hash: 4a4ef493e3a1488c6e321570238084b38948f6db
commit-date: 2026-03-02
host: x86_64-unknown-linux-gnu
release: 1.94.0
LLVM version: 21.1.8
```

**Comment by diksipav** [comment]
I researched this issue and this is what I found: This change [config.opts.unstable_opts.flatten_format_args = false](https://github.com/rust-lang/rust-clippy/commit/89f0aaaec60f5e67604fd2f98b1e43500c1a0578) to keep HIR closer to AST for correct linting introduced this issue with lifetime. `do_fmt2` func from above works with rustc, but Clippy gives false positive.

```
 if self.tcx.sess.opts.unstable_opts.flatten_format_args {
    fmt = flatten_format_args(fmt);
    fmt = self.inline_literals(fmt);
 }
```

I believe that `self.inline_literals(fmt)` is what messed up Clippy lifetimes, when `flatten_format_args` flag is true: 42 is inlined in the template string at compile time and it works. With this flag set to false and not using the let binding no temporary lifetime extension rule applies and borrow checker fails to confirm the lifetime extends to the outer `write_fmt` statement.

Turning `unstable_opts.flatten_format_args ` back on will fix this but will break some ui tests, because now these literal format arguments are not available anymore in HIR and this will break `literal_string_with_formatting_args` and similar lints that inspect format args. Maybe we can tackle this and fall back to check for arguments in AST when they are needed, but not sure if this is the right approach. I used Claude Code to see if I can make this work and I was able to fallback to AST when needed but still I'm not convinced this is the right approach.

These 3 tests fail if we enable back `flatten_format_args` flag:
tests/ui/format.rs
tests/ui/unused_format_specs.rs
tests/ui/useless_concat.rs.

`useless_concat` fails on line 29 [println!("b: {}", concat!("a"))](https://github.com/rust-lang/rust-clippy/blob/master/tests/ui/useless_concat.rs#L29), it should show a warning but it doesn't show it anymore.
`format` fails on line 35 [ format!("{}", "foo")](https://github.com/rust-lang/rust-clippy/blob/master/tests/ui/format.rs#L35) because "foo" gets inlined so this arg does not exist in HIR. It should show a warning but it doesn't show it anymore.

**Comment by oxalica** [comment]
> I researched this issue and this is what I found: This change [config.opts.unstable_opts.flatten_format_args = false](https://github.com/rust-lang/rust-clippy/commit/89f0aaaec60f5e67604fd2f98b1e43500c1a0578) to keep HIR closer to AST for correct linting introduced this issue with lifetime. `do_fmt2` func from above works with rustc, but Clippy gives false positive.

🤦 Alright, so if I use `format!("{}", String::new())` then `do_fmt2` fails in both rustc and clippy. Unfortunately in my use case, these functions are proc-macro-generated and the arguments may or may not be literals. Then I probably should just use `match .. { _ => write!() }` to eliminate all these temporary issues (including https://github.com/rust-lang/rust/issues/145422 which I also ran into)...
This sounds more like a rustc issue to me. I won't expect any optimization to have effects on lifetime checking. I did found that whether `format_args!` would extend temporaries are said to be "unspecified" [in the reference](https://doc.rust-lang.org/stable/reference/expressions.html#r-expr.super-macros.format_args.super-temporaries). So does it mean clippy's behavior is also considered correct?

**Comment by diksipav** [comment]
hm yes, this doesn't work (if you meant that) neither in rustc nor in Clippy, it works with let binding
```
fn do_fmt(w: &mut dyn fmt::Write, mode: u8) -> fmt::Result {
    w.write_fmt(match mode {
        0 => format_args!("foo"),
        1 => format_args!("bar {}", String::from("hej")),
        _ => unreachable!(),
    })
}
```
I still don't understand temporary lifetimes and extensions well enough, so can't provide better answer.

note: hm it is also interesting that when just using `format!("{}", String::new());` itself Clippy gives warning with a suggestion to use `consider using `.to_string()`: `String::new().to_string()` instead of just `String::new()`. @llogiq this should only tell users to just use `String::new()` right?

**Comment by diksipav** [comment]
cc @m-ou-se Hey! I hope it’s okay if I ping you here. From my research, you seem like one of the best people to review this and suggest a solution. :)

**Comment by m-ou-se** [comment]
> This sounds more like a rustc issue to me.

I agree. These lines shouldn't behave differently wrt temporary lifetimes:

```rust
    drop(match () { _ => format_args!("{}", {42}) }); // error (42 not inlined)
    drop(match () { _ => format_args!("{}",  42 ) }); // ok    (42 inlined)
```

I'm not sure if it's worth fixing though.

**Comment by oxalica** [comment]
> > This sounds more like a rustc issue to me.
> 
> I agree. These lines shouldn't behave differently wrt temporary lifetimes:
> 
>     drop(match () { _ => format_args!("{}", {42}) }); // error (42 not inlined)
>     drop(match () { _ => format_args!("{}",  42 ) }); // ok    (42 inlined)
> 
> I'm not sure if it's worth fixing though.

In my mental model, `format_args!` conceptually expands to something like `Arguments { dyns: [&field1, &field2, ...],  ... }` [^1], so both of your examples should compile, ~~`drop(match () { _ => format_args!("{}", String::new()) });` would also compile~~[^2]. It should have nothing to do with inlining at all. How hard would it be to make it behave like this?

The current behavior is so complicated and unpredictable to me. It is definitely not good to have a special case for `format_args!` appearing in Rust reference just to say "unspecified".

[^1]: This simple expansion does not involve super let, and automatically makes `format_args!("{}", IMPL_DISPLAY_CONST)` works in const context. It is like the old `pin!` expansion without `super let`. `super let` would be good to fix the public field problem, but it should just be an explicit lifetime extension, as if using `&` inside struct literals, without subtle different temporary scoping rules.
[^2]: EDIT: My bad. Function arguments are not lifetime extending expression, `drop(match () { _ => &String::new() });` does not compile either. Putting them in `let` initializer should work though.

**Comment by m-ou-se** [comment]
> so both of your examples should compile

Hm, yes, that sounds right.
